### PR TITLE
✨ Introduce version check validation for generate provider command

### LINKED ISSUE #5864: `clusterctl generate provider` should return early if working with incompatible provider versions
**User Story**

As a user when I run `clusterctl generate provider` on a version of a provider that is incompatible I want it to return early with an error messages that says that the version is incompatible with the version of clusterctl.

**Detailed Description**

Ref: https://github.com/kubernetes-sigs/cluster-api/issues/5851#issuecomment-993088847 for more details.

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue.]

/kind feature
/area clusterctl

### PR DESCRIPTION
**Pull request by tharun208**
Signed-off-by: tharun <rajendrantharun@live.com>

This PR introduces validation for incompatible versions while generating providers using the `generate provider` command.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5864

### CODE REVIEW

**Comment by ykakarap** [comment]
@tharun208 Thank you for getting the PR out. 🙂 
I will review this soon.

File: cmd/clusterctl/client/config.go [hunk 1]
```diff
@@ -54,6 +55,45 @@ func (c *clusterctlClient) GetProviderComponents(provider string, providerType c
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Get a client for the provider repository and read the provider components;
+	// during the process, provider components will be processed performing variable substitution, customization of target
+	// namespace etc.
+	// Currently we are not supporting custom yaml processors for the provider
+	// components. So we revert to using the default SimpleYamlProcessor.
+	repositoryClientFactory, err := c.repositoryClientFactory(RepositoryClientFactoryInput{Provider: providerConfig})
```
Comment (ykakarap) [inline cmd/clusterctl/client/config.go:76]:
```suggestion
	repositoryClient, err := c.repositoryClientFactory(RepositoryClientFactoryInput{Provider: providerConfig})
```

nit: the factory returns a client not another factory.

File: cmd/clusterctl/client/config.go [hunk 2]
```diff
@@ -54,6 +55,45 @@ func (c *clusterctlClient) GetProviderComponents(provider string, providerType c
...
+	providerConfig, err := c.configClient.Providers().Get(providerName, providerType)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get a client for the provider repository and read the provider components;
+	// during the process, provider components will be processed performing variable substitution, customization of target
+	// namespace etc.
+	// Currently we are not supporting custom yaml processors for the provider
+	// components. So we revert to using the default SimpleYamlProcessor.
```
Comment (ykakarap) [inline cmd/clusterctl/client/config.go:75]:
```suggestion
```

IMHO, I dont think you need these comments here. Also, If the defaulting changes these comments will become false.

File: cmd/clusterctl/client/config.go [hunk 3]
```diff
@@ -54,6 +55,45 @@ func (c *clusterctlClient) GetProviderComponents(provider string, providerType c
 	return components, nil
 }
 
+func (c *clusterctlClient) GenerateProvider(providerName string, providerType clusterctlv1.ProviderType, options ComponentsOptions) (Components, error) {
+
+	providerName, providerVersion, err := parseProviderName(providerName)
+	if err != nil {
+		return nil, err
```
Comment (ykakarap) [inline cmd/clusterctl/client/config.go:62]:
```suggestion
		return nil, errrors.Wrap(err, "failed to parse provider name")
```

Wrapping the errors provides more context when something fails. 
If you can also wrap other errors below that would be great.

File: cmd/clusterctl/client/config.go [hunk 4]
```diff
@@ -54,6 +55,45 @@ func (c *clusterctlClient) GetProviderComponents(provider string, providerType c
...
+	}
+	// Gets the contract for the current release.
+	currentVersion, err := version.ParseSemantic(providerVersion)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse current version for the %s provider", providerName)
+	}
+	releaseSeries := latestMetadata.GetReleaseSeriesForVersion(currentVersion)
+	if releaseSeries.Contract != clusterv1.GroupVersion.Version {
+		return nil, errors.Errorf("current version of clusterctl is only compatible with %s providers, detected %s for provider %s", clusterv1.GroupVersion.Version, releaseSeries.Contract, providerVersion)
```
Comment (ykakarap) [inline cmd/clusterctl/client/config.go:92]:
```suggestion
		return nil, errors.Errorf("current version of clusterctl is only compatible with %s providers, detected %s for provider version %s", clusterv1.GroupVersion.Version, releaseSeries.Contract, providerVersion)
```

File: cmd/clusterctl/client/config.go [hunk 5]
```diff
@@ -54,6 +55,45 @@ func (c *clusterctlClient) GetProviderComponents(provider string, providerType c
...
+	// during the process, provider components will be processed performing variable substitution, customization of target
+	// namespace etc.
+	// Currently we are not supporting custom yaml processors for the provider
+	// components. So we revert to using the default SimpleYamlProcessor.
+	repositoryClientFactory, err := c.repositoryClientFactory(RepositoryClientFactoryInput{Provider: providerConfig})
+	if err != nil {
+		return nil, err
+	}
+
+	latestMetadata, err := repositoryClientFactory.Metadata(providerVersion).Get()
```
Comment (ykakarap) [inline cmd/clusterctl/client/config.go:81]: This command is also supposed to work when a user does not provider a version in the input (Example: `clusterctl generate provider -i aws`). In such cases the `providerVersion` here will be empty and this command will fail. If the user does not provider a version we have to fall back to the default version.

However, I see that the repository client here does not have a `DefaultVersion` function. I would suggest adding a `DefaultVersion` function to the repository.Client interface and to the implementation (`type repositoryClient struct {...}`).
You can have something like this:
```go
func (c *repositoryClient) DefaultVersion() string {
	return c.repository.DefaultVersion()
}
```
@fabriziopandini Thoughts about adding the `DefaultVersion` function to `repository.Client` interface? This interface already has a `GetVersions` function. I dont think it is unreasonable to add a `DefaultVersion` function.


Then, once you have access to the DefaultVersion function, after l.63 you can have to check to see if `providerVersion` is empty and set it to default version. This should fix the broken use case.

**Review by ykakarap — CHANGES_REQUESTED** [review]
First pass review: 
It mostly looks good, expect for the default version case.

**Comment by tharun208** [comment]
@ykakarap will address the feedbacks this week

**Comment by ykakarap** [comment]
@tharun208 will you have bandwidth to follow up on this?

**Comment by vincepri** [comment]
Any updates on this PR?

**Comment by tharun208** [comment]
@ykakarap can you follow up on this pr?

**Comment by ykakarap** [comment]
> @ykakarap can you follow up on this pr?

I am not sure I understand. If you are asking about the next steps to do then the review on the PR still valid. Adding those changes should be a good place to start.
Let me know if you need help with this.

**Comment by sbueringer** [comment]
@tharun208 Do you have time to address the findings?

**Comment by vincepri** [comment]
Going to close this for now due to inactivity, please feel free to reopen
/close

**Pull request closed without merging**